### ISSUE TITLE
Incorrect CS8625 warning for default parameter values

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#43398]
**Version used:**

```
Branch master (10 Mar 2020)
Latest commit 441c154 by msftbot[bot]:
Merge pull request #42303 from CyrusNajmabadi/useCompoundParens

Support parenthesized expressions in use-compound-operator.
```

**Steps to Reproduce:**

Compile the following code
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;
class C
{
    void M1<T>(T t = default) where T: class? { }
    void M2<T>([AllowNull] T t = default) where T: class? { }
    void M3<T>(T t) where T: class? { t = default; }
}
```

**Expected Behavior:**
No warnings

**Actual Behavior:**
`warning CS8625: Cannot convert null literal to non-nullable reference type.` is reported for `M1` and `M2` but not `M3`

**Notes**
Roslyn doesn't report warnings for null assignments to variables of types that cannot be annotated like `T | where T: class?` so `M3` doesn't have any warnings. Due to the same reasoning both `M1` and `M2` also shouldn't have warnings. It's also possibly to argue that `[AllowNull]` annotation can be used as such syntax so parameters' default values are not the same as regular assignments but even in this case `M2` shouldn't have any warnings

**Comment by RikkiGibson** [comment]
Let's solve this at the same time as #40818

**Comment by RikkiGibson** [comment]
I'm not certain whether this is a bug--the same behavior is present when assigning `default` to unconstrained `T`. /cc @cston

**Comment by TessenR** [comment]
@RikkiGibson 

Are you referring to the warning on `M1? It might be correct actually.

However, I'm still convinced that there should be no warning on `M2` though. The `[AllowNull]` annotation explicitly allows the null value and there's no warning on the argument of the call in the following example:
```
#nullable enable
using System.Diagnostics.CodeAnalysis;
class C
{
  void M2<T>([AllowNull] T t = default) where T : class? => M2<T>(default); // warning on parameter but not on argument
}
```

**Comment by RikkiGibson** [comment]
Revisiting this: we no longer give warnings for any of these, because we just simulate an assignment of the default value to the parameter and join it into the initial state.

We may *start* warning on these once we introduce a warning wave for assigning maybe-null to type parameters which can't be annotated in C# 8. However, this warning should be present in `M1` and `M3`, and not in `M2`.
